### PR TITLE
Fixes #13115 - Adds locations in smart_proxies#index

### PR DESCRIPTION
**Pull request by shlomizadok**
Also changed the status icons to patternfly icons

### CODE REVIEW

**Comment by shlomizadok** [comment]
:scream: `Errno::ENOMEM: Cannot allocate memory - /usr/bin/phantomjs...`

**Comment by dLobatog** [comment]
@shlomizadok Not sure I agree with URLs being 'useless' as the redmine issue describes. I'd rather keep the URL since all objects are already scoped via the top-left taxonomies switcher. All users aside from admins who can set their location to 'Any location' will only see 1 location in the table, whatever is chosen there. 

Aside from that:
- Refresh button is still a glyphicon - patternfly recommends fa-refresh I think (https://www.patternfly.org/styles/icons)

Thanks!

**Comment by domcleal** [comment]
> Not sure I agree with URLs being 'useless' as the redmine issue describes

Indeed, I commented on it: http://projects.theforeman.org/issues/13115#note-3

**Comment by ohadlevy** [comment]
I would argue that url are less important, as by default our installer uses the host fqdn as the name. so by default, you would get: Name:fqdn, url:https://fqdn:8443.

further, smart proxies, since they usually manage infrastructure, are most likely to be bound to location, therefore have the location (and orgs) makes more sense to me then URL.

unlike the media, which the URL is the actual content, in smart proxies, imho URL are less important. granted that they would be displayed in the show page, and can be used for searching.

File: app/assets/javascripts/proxy_status.js [hunk 1]
```diff
@@ -7,11 +7,11 @@ function setItemStatus(item, response) {
   if(response.success) {
     item.attr('title', __('Active'));
     item.addClass('text-success');
-    item.html(icon_text('ok-sign', "", "glyphicon"));
+    item.html(icon_text('ok', "", "pficon"));
```
Comment (ohadlevy) [inline app/assets/javascripts/proxy_status.js:14]: does this change also the about page? IMHO we should change these icons globally (host list, about page, remote execution, scap etc).

Comment (shlomizadok) [inline app/assets/javascripts/proxy_status.js:14]: @ohadlevy - 87 glyphicon occurrences on our code base. this would need a separate pr

Comment (ohadlevy) [inline app/assets/javascripts/proxy_status.js:14]: created http://projects.theforeman.org/issues/13124

**Comment by dLobatog** [comment]
@ohadlevy Check out my comment - the location field in the table will always be the same as whatever you choose on the taxonomies selector. So it'd end up being useless except for admins (who can set 'Any location' on the taxonomies selector.

**Comment by ohadlevy** [comment]
@dLobatog I guess we can display it based on your permissions instead?

**Comment by shlomizadok** [comment]
Just got a tweet from a proxy url. He said he is not useless. I was really touched and reverted.
Adding locations is important because usually a proxy is coupled to location.
If I am an admin, I'll see all proxies with all their locations.
If I am an Org manager, I'll see proxies in several locations.
If I am a Location manager, I'll see all my proxies with the same location

**Comment by ohadlevy** [comment]
what is org/location manager?

**Comment by shlomizadok** [comment]
someone :)
The meaning was a manager / site manager in the context of specific organization or specific location

File: app/views/smart_proxies/index.html.erb [hunk 2]
```diff
@@ -8,7 +8,8 @@
     <tr>
       <th class="col-md-2"><%= s_("SmartProxy|Name") %></th>
       <th class="col-md-4"><%= s_("SmartProxy|Url") %></th>
-      <th class="col-md-4"><%= _("Features") %></th>
+      <th class="col-md-4"><%= _('Locations') %></th>
```
Comment (dLobatog) [inline app/views/smart_proxies/index.html.erb]: That header & column should be hidden when Locations is disabled, otherwise it just shows 'None found'

Comment (shlomizadok) [inline app/views/smart_proxies/index.html.erb]: Added

File: test/integration/smart_proxy_test.rb [hunk 3]
```diff
@@ -3,6 +3,8 @@
 class SmartProxyIntegrationTest < ActionDispatch::IntegrationTest
   test "index page" do
     assert_index_page(smart_proxies_path,"Smart Proxies","New Smart Proxy",false)
+    visit smart_proxies_path
+    assert page.has_selector?('th', :text => 'Locations')
```
Comment (ohadlevy) [inline test/integration/smart_proxy_test.rb]: this will fail if location is disabled.

**Comment by shlomizadok** [comment]
[test]

**Comment by shlomizadok** [comment]
[test]

**Comment by shlomizadok** [comment]
[test]

**Comment by lzap** [comment]
Ok this looks good, it's a good compromise for both upstream and downstream I think. Opinions @dLobatog @domcleal ?

**Comment by domcleal** [comment]
My full comments are on the ticket above - but it looks incomplete to me, as it introduces inconsistencies with orgs and other indexes.

**Comment by lzap** [comment]
Yeah I think adding both Org and Loc is fine along with limiting the output. Maybe combining them both into one column in some reasonable format.

File: app/assets/javascripts/proxy_status.js (hunk 1 above)
Comment (ohadlevy) [inline app/assets/javascripts/proxy_status.js:14]: overlaps with #3083 i would remove it from this pr

**Comment by ohadlevy** [comment]
@lzap its possible to have multi locations/orgs, I cant see how a single  column would work.
@shlomizadok I would suggest to add Orgs too (if its enabled).

**Comment by lzap** [comment]
@ohadlevy I was thinking something like "Org1/Loc1, Org1/Loc2, Org2/Loc3", only to show few initial items making sure it's not wrapped and mouse hover for the rest.

**Comment by lzap** [comment]
The problem with adding both Org and Loc columns is that it takes a lot of horizontal space. And it looks like we are making a precedent here - other pages might get the same too. It could be enough space on this index page, but not on others. That's why I was thinking about some combined column as the code needs to iterate thorugh multiple values anyway.

Combined output is only problematis when user has one name for both org and loc. This would be problem on a detail page, but since this is just an index page which aims to provide shortest and clearest overview, I think this could be enough.

We could also leverage a separator like slash. An example of few combinations:
- Org1 / Loc1
- Org1, Org2 / Loc1
- Org1 / Loc1, Loc2
- Org1, Org2 / Loc1, Loc2

The code should also solve edge cases:
- Org1, Org2 / All locations

An alternative approach is to show the taxonomy as you hover through the table, so it is initially invisible. This solves the issue with free space for all our pages, but it is less obvious.

**Comment by lzap** [comment]
Maybe we could add taxonomy information after the name in this form:
- smart-proxy.org.blah.prov.lan (Org1 / Loc1)
- smart-proxy.org.blah.prov.lan (Org1, Org2 / Loc1)
- smart-proxy.org.blah.prov.lan (Org1 / Loc1, Loc2)
- smart-proxy.org.blah.prov.lan (Org1, Org2 / Loc1, Loc2)

This saves space, but apparently you can't sort by locations. But this gives us better flexibility to hide those when Organization and/or Location is selected via the top-left selector. No need to hide a column and break the layout, we can just hide those in the name column.

**Comment by lzap** [comment]
@shlomizadok these are just ideas for dicussion, feel free to simply add an optional Organization column and deal with multiple values. I think for the future we need to think a generic solution for all relevant index pages. Something we don't want to solve in this PR.

**Comment by lzap** [comment]
Moved the discussion to https://groups.google.com/forum/?fromgroups#!topic/foreman-dev/UH-ycYuDEZk and I think we can move on with the simplified implementation. We use the same in Discovery plugin (two simple columns, visible when taxonomy enabled).

**Comment by ohadlevy** [comment]
@shlomizadok please rebase please.

**Comment by shlomizadok** [comment]
Rebased + Added Organizations. It now looks like:
image

File: app/views/smart_proxies/index.html.erb [hunk 4]
```diff
@@ -8,7 +8,13 @@
     <tr>
       <th class="col-md-2"><%= s_("SmartProxy|Name") %></th>
       <th class="col-md-4"><%= s_("SmartProxy|Url") %></th>
-      <th class="col-md-4"><%= _("Features") %></th>
+      <% if SETTINGS[:locations_enabled] %>
+        <th class="col-md-4"><%= _('Locations') %></th>
```
Comment (ohadlevy) [inline app/views/smart_proxies/index.html.erb]: consider using percentage ? @tbrisker ?

File: app/views/smart_proxies/index.html.erb [hunk 5]
```diff
@@ -8,7 +8,13 @@
     <tr>
       <th class="col-md-2"><%= s_("SmartProxy|Name") %></th>
       <th class="col-md-4"><%= s_("SmartProxy|Url") %></th>
```
Comment (ohadlevy) [inline app/views/smart_proxies/index.html.erb]: I would remove the URL based on the fact that our installer create the entry with the same name and url.

**Comment by shlomizadok** [comment]
image
w/o url

**Comment by dLobatog** [comment]
@shlomizadok Seems like the integration test fails on 1.9.3, take a look at it.

**Comment by dLobatog** [comment]
Looks good to me, code-wise, ready to merge after fixing the problem with 1.9.3 and the integration test. I'll leave the merge to you @lzap as you assigned this PR to yourself. I still think URL is more useful than taxonomies though...

Since this came kind of out of the blue (I don't recall any discussion about this anywhere) - @ohadlevy @shlomizadok are there plans to add these columns to other objects, or making what @lzap said, some system to choose what columns to show on each page? 

If we go that route, I'd rather use some pre-baked solution rather than having to maintain it ourselves

**Comment by shlomizadok** [comment]
@dLobatog - Removed url assertion from the integration test, as url is no longer available on the index page. afaik, no plans to add this to other objects. We've added this here because of the unique connection between proxies and locations. Thanks for reviewing!

**Comment by shlomizadok** [comment]
@dLobatog @lzap - tests are :green_apple: 
Please let me know if this plays well or are there other comments needs to be addressed. Thanks :)

File: app/views/smart_proxies/index.html.erb [hunk 6]
```diff
@@ -7,8 +7,13 @@
...
-      <th class="col-md-4"><%= _("Features") %></th>
+      <% if SETTINGS[:locations_enabled] %>
+        <th class="col-md-2"><%= _('Locations') %></th>
+      <% end %>
+      <% if SETTINGS[:organizations_enabled] %>
+        <th class="col-md-2"><%= _('Organizations') %></th>
+      <% end %>
+      <th class="col-md-4"><%= _('Features') %></th>
       <th class="col-md-1"><%= _('Status') %></th>
       <th><%= _('Actions') %></th>
```
Comment (lzap) [inline app/views/smart_proxies/index.html.erb]: Can you add hiding for smaller screens? In my case (Logs) I did this:

``` html
      <tr>
          <th class="ca col-md-2 hidden-xs nbsp">Time</th>
          <th class="ca col-md-1 hidden-sm hidden-xs nbsp">Level</th>
          <th class="col-md-9 col-sm-10 col-xs-12 nbsp">Message</th>
          <th class="hidden"></th>
      </tr>
```

So level and then even timestamp disappears when you make the browser narrower.

File: app/views/smart_proxies/index.html.erb [hunk 7]
```diff
@@ -17,7 +22,12 @@
     <% for proxy in @smart_proxies %>
       <tr class="proxy-show" data-url="<%= ping_smart_proxy_path(proxy) %>">
         <td><%= link_to_if_authorized proxy.name, hash_for_smart_proxy_path(:id => proxy).merge(:auth_object => proxy, :authorizer => authorizer) %></td>
-        <td><%= proxy.url %></td>
+        <% if SETTINGS[:locations_enabled] %>
+          <td><%= generate_links_for(proxy.locations).html_safe %></td>
+        <% end %>
+        <% if SETTINGS[:organizations_enabled] %>
+          <td><%= generate_links_for(proxy.organizations).html_safe %></td>
```
Comment (lzap) [inline app/views/smart_proxies/index.html.erb]: @dLobatog have you tested this? How it looks like when there are many locations/orgs?

I think we need "ellipsis" here so it does not overflow perhaps?

**Comment by lzap** [comment]
There were concerns about applying taxonomy to all pages, that's something we should be tracking (and there is a ticket already in that regard) but it's out of scope for this Smart Proxy improvements. LGTM!

File: app/views/smart_proxies/index.html.erb (hunk 7 above)
Comment (lzap) [inline app/views/smart_proxies/index.html.erb]: Ok this overflows, please add `ellipsis` thing.

**Comment by lzap** [comment]
So only the two usability comments and it is ready to go.

**Comment by shlomizadok** [comment]
Added

File: app/views/smart_proxies/index.html.erb [hunk 8]
```diff
@@ -7,8 +7,13 @@
...
-      <th class="col-md-4"><%= _("Features") %></th>
+      <% if SETTINGS[:locations_enabled] %>
+        <th class="col-md-2 hidden-sm hidden-xs nbsp"><%= _('Locations') %></th>
+      <% end %>
+      <% if SETTINGS[:organizations_enabled] %>
+        <th class="col-md-2 hidden-sm hidden-xs nbsp"><%= _('Organizations') %></th>
+      <% end %>
+      <th class="col-md-4"><%= _('Features') %></th>
       <th class="col-md-1"><%= _('Status') %></th>
       <th><%= _('Actions') %></th>
```
Comment (tbrisker) [inline app/views/smart_proxies/index.html.erb]: Please add a width to the action column so it doesn't expand too much when locs/orgs are disabled

**Comment by tbrisker** [comment]
I'm getting an N+1 here, other then that :+1:

**Comment by shlomizadok** [comment]
@tbrisker - Rails doesn't play well with polymorphic includes (fixed at Rails 5).
At the moment, we can reduce N+1 on only one polymorphic association (either organization or location) - but sadly not to both of them. Please advise

**Comment by shlomizadok** [comment]
Sending now with fix of N+1

File: app/controllers/smart_proxies_controller.rb [hunk 9]
```diff
@@ -108,4 +108,11 @@ def action_permission
         super
     end
   end
+
+  def proxy_resource_base
```
Comment (shlomizadok) [inline app/controllers/smart_proxies_controller.rb]: Not sure I like it this way, what says you?

Comment (lzap) [inline app/controllers/smart_proxies_controller.rb]: I am okay with that, we do similar things around the codebase, this one is three lines but what can we do about it. @tbrisker ? Please ack and I am gonnat give it a whirl.

Comment (tbrisker) [inline app/controllers/smart_proxies_controller.rb]: How about something along the lines of overriding `resource_base` and using super, something like:

```
base = super
base = base.eager_load(:locations) if SETTINGS[:locations_enabled]
base = base.eager_load(:organizations) if SETTINGS[:organizations_enabled]
base
```

Comment (shlomizadok) [inline app/controllers/smart_proxies_controller.rb]: done

**Comment by tbrisker** [comment]
:shipit: when tests are :green_apple:

**Comment by lzap** [comment]
Looks like there is plenty of space on LG devices, could you do one additional change?

image

Can you please keep the URL column and show it on large (LG) screen only? For medium let's just hide it.

As an additional bonus, you could add the URL as a tooltip to the name. But that's not required, just an idea. Other than that GTM!

**Comment by shlomizadok** [comment]
Yeah, I'll add it so someone else will tell me to remove it.

**Comment by shlomizadok** [comment]
Also, I don't see much space to add the url column

**Comment by tbrisker** [comment]
I don't thing we should add the url field back, putting it in a tooltip might be nice but I think we should just leave it at that for now and look into a tooltip if required in the future.

**Comment by shlomizadok** [comment]
Don't think we need a tooltip - the foreman installer is naming the proxy by its name so it's a dup content

**Comment by lzap** [comment]
The problem I am trying to solve with my proposal appears when ORG/LOC are disabled, then the page renders just four columns and on the large display there is _plenty_ of space, really. And then it looks like we removed the URL for nothing in this scenario. That's what I don't like, and it's not just me here. And that's why I asked if you could do this change.

If you think there is no space for it when ORG/LOC is enabled, that's a fair point. Consider showing it only when ORG/LOC is/are disabled and only on large screen.

I don't think we should resign at this point saying "URL is not always relevant" and hiding it even for users who don't care about taxonomy. And there are plenty of them! Come on, I am not asking much. We are talking about three lines change.

**Comment by shlomizadok** [comment]
With locations / organizations enabled it looks like:
image
Does this look reasonable? (It has a tooltip to display the full url)

Without locations / organizations it looks like:
image

With one of the enabled:
image

If this looks okay, I'll send the change. If not - please advise

**Comment by lzap** [comment]
To me it looks like all is good, except the first example. I'd hide URL in case both ORG/LOC are present. That was my last comment. If you move the hover tip to name, then you don't need to take care about earger loading...

Let me repeat what I said to be clear: I totally agree with ditching URL _when ORG/LOC are both enabled_, but not when one or none is present.

**Comment by shlomizadok** [comment]
Added. With both locations & organizations it looks like:
image

With one enabled, or with none it looks like the above.

**Comment by lzap** [comment]
One nitpick - could you also add "ellipsis" to Features cells? Because in my testing, I have so many features so it breaks and it's not nice:

image

**Comment by shlomizadok** [comment]
Added

**Comment by lzap** [comment]
Merged as b8f6567e23ec2f4d9c14768b67d5a044725bc89d, thank you!

**Pull request closed without merging**